astropy/io/ascii/ui.py

```diff
@@ -460,12 +471,19 @@ def _guess(table, read_kwargs, format, fast_reader):
     # keep track of the failed guess and move on.
     for guess_kwargs in filtered_guess_kwargs:
         t0 = time.time()
+        # update guess_kwargs again; need a deep copy to preserve dicts
+        for key, val in user_kwargs.items():
+            if key not in guess_kwargs:
+                guess_kwargs[key] = copy.deepcopy(val)
+            elif val != guess_kwargs[key] and guess_kwargs != fast_kwargs:
+                guess_kwargs[key] = copy.deepcopy(val)
```

Comment:
Apparently necessary since otherwise, after the first fast reader returned by `_get_guess_kwargs_list` (usually `fastbasic.FastBasic`) was (unsuccessfully) tried, the `fast_kwargs` were lost.

Reply:
Right, and it would make for cleaner code design. As it turns out, at the core of the issue is not just `kwargs` dicts being modified, but more importantly that `_get_guess_kwargs_list` never sets them in the first place for most of the guesses - only for the `FixedWidthTwoLine` and `*Basic` ones at this point.
The current guessing order is thus:
```
html.HTML, ecsv.ECSV,
```
without (i.e. ignoring) any `kwargs`;
```
fixedwidth.FixedWidthTwoLine,
fastbasic.FastBasic, basic.Basic,
```
with `kwargs`, if any were supplied;
```
fixedwidth.FixedWidthTwoLine,
fastbasic.FastBasic, basic.Basic,
basic.Rdb,
fastbasic.FastTab, basic.Tab,
cds.Cds,
daophot.Daophot,
sextractor.SExtractor,
ipac.Ipac,
latex.Latex, latex.AASTex,
```
without any `kwargs`, and finally
```
fastbasic.FastCommentedHeader, basic.CommentedHeader,
fastbasic.FastBasic, basic.Basic,
fastbasic.FastNoHeader, basic.NoHeader
```
with all combinations of `delimiter` and `quotechar`, but again ignoring any `read_kwargs`.

This makes it almost a design decision now - is this particular handling and non-handling of user-supplied arguments intended? IMO I would expect `_guess` to try to use all arguments explicitly given by the user, with all (and only the) readers that support them.
In any case there is no test in `io.ascii` that expects otherwise...